### PR TITLE
Fix formatting

### PR DESCRIPTION
**Pull request by bryanforbes**
### Description
Updates code style and pre-commit checks to match sqlalchemy. A few things to note:

* `# fmt: off` + `# fmt: on` was needed to keep black and zimports from fighting
* I added `flake8-pyi` and did not include `flake8-builtins`, `flake8-docstrings`, and `flake8-rst-docstrings` because they are irrelevant for `pyi` files

### Checklist

This pull request is:

- [ ] A documentation / typographical error fix
	- Good to go, no issue or tests are needed
- [x] A short code fix
	- please include the issue number, and create an issue if none exists, which
	  must include a complete example of the issue.  one line code fixes without an
	  issue and demonstration will not be accepted.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

### CODE REVIEW

**Comment by CaselIT** [comment]
> # fmt: off + # fmt: on was needed to keep black and zimports from fighting

Interesting. Sqlalchemy uses both and they are nice to each other. What's the issue?

**Comment by zzzeek** [comment]
> > `# fmt: off + # fmt: on` was needed to keep black and zimports from fighting
> 
> Interesting. Sqlalchemy uses both and they are nice to each other. What's the issue?

zimports doesn't really work for .pyi files because there are special meanings to a statement like:

   from .foo import Foo as Foo

vs.

  from .foo import Foo

zimports doesn't know about that.   I've removed zimports from the .pre-commit-config.yml for sqlalchemy2 so there should be no need to worry about zimports for now.

**Comment by bryanforbes** [comment]
> > `# fmt: off + # fmt: on` was needed to keep black and zimports from fighting
> 
> Interesting. Sqlalchemy uses both and they are nice to each other. What's the issue?

The issue is that `black` wants to take long imports such as:

```python
from .langhelpers import monkeypatch_proxied_specials as monkeypatch_proxied_specials
```

Into:

```python
from .langhelpers import (
    monkeypatch_proxied_specials as monkeypatch_proxied_specials,
)
```

And `zimports` turns the `black` output back into:

```python
from .langhelpers import monkeypatch_proxied_specials as monkeypatch_proxied_specials
```

The `# fmt: off` / `# fmt: on` switch tells `black` to leave the imports alone and let `zimports` handle it.

> zimports doesn't really work for .pyi files because there are special meanings to a statement like:

It doesn't do too bad of a job, to be honest. IMO, it would be better to keep it on and match the import format of `sqlalchemy`.

**Comment by zzzeek** [comment]
> > > `# fmt: off + # fmt: on` was needed to keep black and zimports from fighting
> > 
> > 
> > Interesting. Sqlalchemy uses both and they are nice to each other. What's the issue?
> 
> The issue is that `black` wants to take long imports such as:
> 
> ```python
> from .langhelpers import monkeypatch_proxied_specials as monkeypatch_proxied_specials
> ```
> 
> Into:
> 
> ```python
> from .langhelpers import (
>     monkeypatch_proxied_specials as monkeypatch_proxied_specials,
> )
> ```
> 
> And `zimports` turns the `black` output back into:
> 
> ```python
> from .langhelpers import monkeypatch_proxied_specials as monkeypatch_proxied_specials
> ```
> 
> The `# fmt: off` / `# fmt: on` switch tells `black` to leave the imports alone and let `zimports` handle it.
> 
> > zimports doesn't really work for .pyi files because there are special meanings to a statement like:
> 
> It doesn't do too bad of a job, to be honest. IMO, it would be better to keep it on and match the import format of `sqlalchemy`.

I was hitting the same problems as you and was not able to commit, which is why it's not in pre-commit right now.   I'm not sure how to commit w/o disabling checks if zimports is in pre-commit.    overall I dont have time to figure it out.

**Comment by bryanforbes** [comment]
> I was hitting the same problems as you and was not able to commit, which is why it's not in pre-commit right now. I'm not sure how to commit w/o disabling checks if zimports is in pre-commit. overall I dont have time to figure it out.

Sorry if I wasn't clear: using the `# fmt: off` / `# fmt: on` trick lets `zimports` alone handle the formatting so `pre-commit` works. In other words, I figured it out for you.

**Comment by CaselIT** [comment]
> from .foo import Foo as Foo
>
> vs.
>
> from .foo import Foo

oh, what's the difference? Sorry but I'm still a noob on python stubs and mypy.

> The issue is that black wants to take long imports such as: ...

the issue seems to be the line length then. @zzzeek thoughts on using 89 or 100 for the stubs?

**Comment by layday** [comment]
> oh, what's the difference? Sorry but I'm still a noob on python stubs and mypy.

`Foo as Foo` means that `Foo` is public and can be imported from other modules.  This is to distinguish between symbols you import for annotation purposes and symbols you genuinely want to export as part of your API.  mypy in strict mode will emit an error if you attempt to import a symbol which isn't declared as being public, either by adding a synonymous 'as' clause or by referencing the symbol in `__all__`.

**Comment by CaselIT** [comment]
Thanks! I guessed it was something like this but I wanted to be sure.

At first thought it's a strange convention, since it's the opposite of pythons (one usually use `from .foo import Foo as _Foo` to make things private). But I guess we will have to adapt :)

**Comment by bryanforbes** [comment]
> > from .foo import Foo as Foo
> > vs.
> > from .foo import Foo
> 
> oh, what's the difference? Sorry but I'm still a noob on python stubs and mypy.

No need to apologize. Stub files only re-export a subset of the symbols that are in the scope of the file. [The stub files section of PEP 484](https://www.python.org/dev/peps/pep-0484/#stub-files) does a pretty good job of explaining it, but the basics are:

* All symbols that do not begin with `_` are re-exported except imports.
* Only imports that are imported with `import X as X` or `from Y import X as X` will be re-exported.

**Comment by zzzeek** [comment]
> > I was hitting the same problems as you and was not able to commit, which is why it's not in pre-commit right now. I'm not sure how to commit w/o disabling checks if zimports is in pre-commit. overall I dont have time to figure it out.
> 
> Sorry if I wasn't clear: using the `# fmt: off` / `# fmt: on` trick lets `zimports` alone handle the formatting so `pre-commit` works. In other words, I figured it out for you.

oh, OK using fmt on/off, sure.    That's fine.  thanks!    do folks here want to be part of the sqlalchemy2-stubs project for now and I can just grant commit access?   this is all new stuff that just needs to get done.

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -8,4 +8,22 @@ repos:
...
         files: \.(py|pyi)$
 
+-   repo: https://github.com/sqlalchemyorg/zimports
+    rev: 0.3.0
+    hooks:
+    -   id: zimports
+        types: [file]
+        files: \.(py|pyi)$
+
+-   repo: https://github.com/pycqa/flake8
```
Comment (zzzeek) [inline .pre-commit-config.yaml:18]: i had problems getting flake8 to work too, but if you know how to keep it working then great!

**Comment by CaselIT** [comment]
> do folks here want to be part of the sqlalchemy2-stubs project for now and I can just grant commit access? this is all new stuff that just needs to get done.

@layday we also have a gitter chat at https://gitter.im/sqlalchemy/devel if you want to talks things

**Comment by zzzeek** [comment]
OK I read the PR now, it looks great.     There's no CI for the stubs right now except for when they are merged to master, so I will just go for it.

File: .pre-commit-config.yaml (hunk 1 above)
Comment (bryanforbes) [reply .pre-commit-config.yaml:18]: Yeah, I made sure to get everything passing before I created the PR :)

**Comment by CaselIT** [comment]
> 
> 
> OK I read the PR now, it looks great. There's no CI for the stubs right now except for when they are merged to master, so I will just go for it.

if you have some commands to run to test the pr I can add one here on github. But at the moment I don't know what would need to run

**Comment by zzzeek** [comment]
yeah how can we "test" these anyway?    I've noticed that mypy just ignores stubs that have obvious mistakes like missing imports.  whats the right way to test things like that?

**Comment by bryanforbes** [comment]
> yeah how can we "test" these anyway?

Three ways:

* Run `mypy` on the stub files themselves to ensure they are valid stub files (we'd need to do this for all supported versions, including 2.7)
* We can use something like [pytest-mypy-plugins](https://github.com/typeddjango/pytest-mypy-plugins) to ensure that the stubs and plugin behave correctly (ex. a method returning a generic type returns the correct type)
* Use `stubtest` to ensure all public methods/properties are covered by the stubs

**Pull request merged**